mltensor-explainer.md

```diff
@@ -0,0 +1,421 @@
+const mlTensor2 = await mlContext.createTensor({..., usage: MLTensorUsage.WEBGPU_INTEROP});
+
+const applyEffectToFrame = async () => {
+  const gpuVideoTexture = gpuDevice.importExternalTexture({source: video});
+
+  // Rent out the MLTensor to WebGPU.  
+  const tensorizedGpuBuffer = gpuDevice.importExternalBuffer(mlTensor1);
+
+  // Create a bind group for `gpuVideoTexture`, create a command encoder, etc.
+  // to "tensorize" `gpuVideoTexture` and store the result in `tensorizedGpuBuffer`
+  // ...
+
+  gpuDevice.queue.submit([tensorizationCommandEncoder.finish()]);
+
+  // Return the buffer to WebNN.
+  tensorizedGpuBuffer.destroy();
+
+  // Perform some inference described by `graph` on the frame
+  // (e.g. selfie segmentation)
+  mlContext.dispatch(
+    graph,
+    /*inputs=*/{'input': mlTensor1},
+    /*outputs=*/{'output': mlTensor2},
+  );
+
+  // Rent the other MLTensor out to WebGPU.
+  const tensorizedGpuBufferAfterInference = gpuDevice.importExternalBuffer(mlTensor2);
+
+  // Create a bind group for `tensorizedGpuBufferAfterInference`,
+  // create a command encoder, etc to feed `tensorizedGpuBufferAfterInference`
+  // into a GPU shader which may blur the frame or replace background sections
+  // and then render the result
+  // ...
+
+  gpuDevice.queue.submit([texturizeAndRenderCommandEncoder.finish()]);
+
+  // Return the buffer to WebNN for the next frame.
+  tensorizedGpuBufferAfterInference.destroy();
+
+  // Call this method for each frame.
+  video.requestVideoFrameCallback(applyEffectToFrame);
+}
+```
+
+## Design Discussion
+
+### Timelines
+
+WebNN uses a programming model similar to [WebGPU's](https://www.w3.org/TR/webgpu/#programming-model), in that compute tasks are posted to a timeline - which I've referred to as an "ML context timeline" throughout this document - separate from the content timeline (i.e. "script"). See [the WebGPU documentation of timelines](https://gpuweb.github.io/gpuweb/#programming-model-timelines) for more details.
+
+Specifying WebNN timelines is tracked in [#529](https://github.com/webmachinelearning/webnn/issues/529).
+
+### Device Affinity and Relationship to a `GPUDevice`
+
+The user agent decides where the memory backing an `MLTensor` is allocated. The WebNN API allows the developer to provide hints - primarily via `MLTensorUsageFlags` - but these are not binding.
+
+For example [an `MLContext` may be created with a `GPUDevice`](https://www.w3.org/TR/webnn/#dom-ml-createcontext-gpudevice), and creating an `MLTensor` from this context with the `MLTensorUsage.WEBGPU_INTEROP` flag expresses a clear intention to share the tensor with the given `GPUDevice`. However, there is no guarantee that sharing this tensor with WebGPU will be zero-copy.
+
+The `MLTensorUsage.READ_FROM` and `MLTensorUsage.WRITE_TO` flags likewise are hints to the user agent indicating that the underlying data will be read and written to, respectively, by script.
```

Comment:
Ah, you're correct that usages are required from the JS API's perspective (we'll throw `TypeError` if attempting to misuse an `MLTensor`), but this text is speaking to user agents, not web developers. From the perspective of the user agent, it can allocate the buffer wherever it wants, but we're recommending it use `MLTensorUsage` as a hint in that decision

I updated the first paragraph of this section to hopefully make this more clear. WDYT?

Reply:
The user agent still needs to figure out which device the buffer needs to be allocated on where omitting any `MLTensorUsage` flag results could result into the buffer being left inaccessible (aka only used for output). Are you saying the user agent needs to either predict this and/or move/copy it around as needed?

Reply:
I believe the problem you're describing relates to this question? https://github.com/webmachinelearning/webnn/pull/754/files#diff-0ee55cff3e7c8d2280b48fa37962505a398c7a18df2eb78e0497c76042054cc8R286

Reply:
Sounds like the same problem. Also, if we eliminate `deviceType` then we're back to using hints here so I suppose we need to think about how this will work.

Reply:
Yeah the current assumption is that scoping an `MLTensor` to an `MLContext` is sufficient to determine how to allocate the buffer, but if an `MLContext` is not as closely tied to a compute device (e.g. #749) then that assumption may no longer hold.

I'm not aware of a viable alternative, however, especially while DML does not (yet) have fallback/multi-device support

Reply:
WDYT about going with the (original) assumption that these usage hints are requirements, allocation must occur on `deviceType`, and leaving a TODO for us to figure out how best to convert this into hints?

Reply:
> the (original) assumption that these usage hints are requirements

For the user agent or for script?

I'm proposing that usages are _required and enforced from script_, as originally discussed. But the specification cannot force the user agent to allocate memory on any particular compute unit, just like how WebGPU can't force the system to use a "real" GPU (e.g. Warp exists)

There's a give-and-take between the abstract nature of a spec and the nitty-gritty details of its implementations. We don't want the spec to prescribe implementation details, but of course the spec is written with implementations in mind and accommodates their constraints. If the user agent doesn't have enough information to make a reasonable decision, then we can change the spec.

In this case, I don't think we know (or at least _I don't know_) whether the user agent can make a reasonable decision in the "non-UMA without `deviceType`" scenario described above. I also don't know of any spec changes which would help the user agent make a better decision, in part because I don't know what "non-UMA without `deviceType`" will look like in practice.

If "multi-device" means "one of several compute units is selected during graph compilation" then we could e.g. scope an `MLTensor` to an `MLGraph` rather than an `MLContext` (at least in that scenario). But even that doesn't match a world in which a compiled `MLGraph` is _also_ not tied to a compute unit, as @mwyrzykowski alludes to in #749:

> ... if the WebGPU workload is falling behind the requested update rate, the browser implementation may move the model processing to the device least in use.

...since this is a theoretical concern for if/when DML supports multi-device, this explainer punts on finding a solution for this use case for now ¯\\_(ツ)_/¯

Reply:
Correct, for the user agent it would stay "required" (for now).

> just like how WebGPU can't force the system to use a "real" GPU (e.g. Warp exists)

AFAIK WebGPU doesn't allow a real GPU buffer to be created then used with a WARP device: a buffer must be created from the same device type, even if it happens to be CPU.

> this explainer punts on finding a solution for this use case for now ¯\(ツ)/¯

SGTM but currently, the explainer suggests this is possible as "hints" (ie. omitting flags == stays on `deviceType`) which I think is too speculative.

Reply:
> As I mentioned above, the specification cannot dictate where/how the user agent allocates memory. 

The specification might need to specify where the allocation starts or have additional usages which describe how often the contents will be updated. I think it's doable but not with the `usages` we have mentioned in this explainer.

Reply:
> As I mentioned above, the specification cannot dictate where/how the user agent allocates memory.

We could specify new hints which describe how frequently the tensor's data will change which could help (but not dictate) the user agent's decision where best to allocate it. Another option is we specify the allocation will _start_ on the CPU only. Giving web developers no control over these allocations doesn't seem ideal and we'll probably need to help them there, beyond what this explainer offers.

If you want to address this later, I'm OK with that too.

Reply:
> We could specify new hints which describe how frequently the tensor's data will change which could help (but not dictate) the user agent's decision where best to allocate it

Right, if it was useful to signal "this buffer will be used for WebGPU interop repeatedly and just read back to script once" or vice-versa, then the user agent may be able to make a more informed decision between the choices presented in https://github.com/webmachinelearning/webnn/pull/754#discussion_r1755379335:
> * Allocate the `MLTensor` as a "default" buffer, requiring an additional staging buffer for `readBuffer()`, or
> * Allocate the `MLTensor` as a "readback" buffer and whip up a new GPU buffer for interop

Reply:
> then the user agent may be able to make a more informed decision between the choices presented in https://github.com/webmachinelearning/webnn/pull/754#discussion_r1755379335

The decision is only good as its last access and since inputs could be re-dispatched as outputs, I worry hints specified at creation could lead to bad choices being made repeatedly.

I think it's fine we have hints to determine what type of CPU allocation we start with (eg. read-back or upload buffer) but we'll also need to move it over to the GPU (eg. default buffer) then release the CPU allocation. If the web developer needs to `WRITE` or `READ` again then it'll move back to the CPU. If specified as such, web developers might have enough predictability to avoid staging buffers (and going OOM).